inception-imls-external/src/main/java/de/tudarmstadt/ukp/inception/recommendation/imls/external/ExternalRecommender.java

```diff
@@ -186,6 +207,28 @@ private String serializeCas(CAS aCas) throws RecommendationException
         }
     }
 
+    private Document buildDocument(CAS aCas) throws RecommendationException
+    {
+        CASMetadata casMetadata = getCasMetadata(aCas);
+
+        Document document = new Document();
+        document.setXmi(serializeCas(aCas));
+        document.setDocumentId(casMetadata.getSourceDocumentId());
+        document.setUserId(casMetadata.getUsername());
+
+        return document;
+    }
+
+    private CASMetadata getCasMetadata(CAS aCas) throws RecommendationException
+    {
+        try {
+            return JCasUtil.selectSingle(aCas.getJCas(), CASMetadata.class);
+        } catch (CASException | IllegalArgumentException e) {
+            LOG.error("Error while reading CAS metadata!", e);
```

Comment:
If we are throwing the exception, there is no need to log - we're bound to catch the exception further up and log/display it to the user - this causes annoying multiple loggings of the same exception.

 Also - I don't see why an IOException is a recommendation exception here. I would expect that a recommendation exception is one that is related to the recommendation logic, not one that is thrown in recommender code.

Reply:
The problem is that the interface throws `RecommendationException`, so I have to wrap it.

Reply:
`IllegalArgumentException` is unchecked - doesn't have to be wrapped.

`getCasMetadata()` is a private method - it doesn't implement an interface?

Reply:
The method is called in `predict` and `train`, which are part of the interface. I rather wrap it here than to duplicate the wrapping twice.